File: 133. Clone Graph/main.c
#include <stdio.h>

struct Node {
    int val;
    int numNeighbors;
    struct Node **neighbors;
};

#define ARRAY_SIZE 101

static struct Node *dfs(struct Node *s, struct Node **visited) {
    if (visited[s->val] != NULL) {
        return visited[s->val];
    }

    struct Node *node = (struct Node *)malloc(sizeof(struct Node));
    visited[s->val] = node;
    node->val = s->val;
    node->numNeighbors = s->numNeighbors;

    if (node->numNeighbors == 0) {
        node->neighbors = NULL;
        return node;
    }

    node->neighbors = (struct Node **)calloc(node->numNeighbors, sizeof(struct Node *));
    for (int i = 0; i < node->numNeighbors; i++) {
        node->neighbors[i] = dfs(s->neighbors[i], visited);
    }

    return node;
}

/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     int numNeighbors;
 *     struct Node** neighbors;
 * };
 */
struct Node *cloneGraph(struct Node *s) {
    if (s == NULL) {
        return NULL;
    }

    struct Node *visited[ARRAY_SIZE] = {NULL};
    return dfs(s, visited);
}

File: 133. Clone Graph/main.py
from typing import Optional


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if node is None:
            return None

        visited = dict()

        def dfs(node: Optional["Node"]) -> Optional["Node"]:
            nonlocal visited

            if node in visited:
                return visited[node]

            visited[node] = Node(node.val)

            for neighbor in node.neighbors:
                visited[node].neighbors.append(dfs(neighbor))

            return visited[node]

        return dfs(node)
